feat: support exp manager checkpointing with object store via Multi-Storage Client

jayya2 opened the pull request:
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add support for using [multi-storage-client](https://github.com/NVIDIA/multi-storage-client) for the Exp Manager checkpointing mechanism so that users who haven't migrated to new checkpoints can still take advantage of the object stores.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

titu1994 commented on nemo/core/connectors/save_restore_connector.py:
```diff
@@ -671,12 +689,36 @@ def _tar_open(path2file: str) -> Generator[tarfile.TarFile, None, None]:
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            if is_msc_url:
+                filename_with_extension = path2file.split("/")[-1]  # get the filename with extension
```
Please refactor the code to be similar to older version using _tar_open(). Msc functions should be moved to their own functions, and just called from here rather than embed them inside each function.

jayya2 replied:
fixed, added if else check for msc url and have a separate function tailored for msc calls

titu1994 commented on nemo/utils/callbacks/nemo_model_checkpoint.py:
```diff
@@ -508,7 +527,10 @@ def file_exists(
         self, filepath: str, trainer: "lightning.pytorch.Trainer", check_dist_ckpt: bool = True  # noqa: F821
     ) -> bool:
         """Checks if a file or a file without a suffix (distributed checkpoint) exists."""
-        exists = self._fs.exists(filepath) or (check_dist_ckpt and self._fs.exists(ckpt_to_dir(filepath)))
```
Can you revert this and just add another line to chain your new check? This check is unreadable.

jayya2 replied:
fixed

titu1994 reviewed (commented):
Overall ok, but the code is getting super complicated. 
Can you avoid trying to add is_msc_* to every single check and simply separate the code paths to be completely independent of each other if msc is there or not?

Right now this mix match of merging msc logic with older logic makes the code unreadable.

jayya2 commented:
@titu1994 
thanks for the quick review, I addressed most of the comments:
- Have isolated function to wrap up MSC calls to prevent a mix of logic
- Reduce the number of MSC url checks
- use full name for msc (after second thought I don't want to mix both `multiplestorageclient` and `msc` so I would rather have long name everywhere)

Let me know if further refactor is required, thanks!

titu1994 reviewed (dismissed):
Much cleaner than before, thank you !

ko3n1g commented:
@jayya2 @titu1994 : This PR got affected by multiple infra issues. Please don't relabel this PR, we will restart failed jobs for you and try to get this merged with as little overhead as possible. TIA!

jayya2 commented:
@ericharper @titu1994 @ko3n1g are the infra issues resolved?  saw some of the tests retried and passed but still 2 failing but seems unrelated to this change

pradeep-mj commented on nemo/utils/callbacks/nemo_model_checkpoint.py:
```diff
@@ -644,32 +676,45 @@ def _remove_unfinished_checkpoints(checkpoint_dir: Union[Path, str]) -> None:
+            MULTISTORAGECLIENT_PROTOCOL
+        )
+
+        # TODO: add multistorageclient support for distributed checkpointing
```
Should we call this as not supported instead?
